src/job_metadata.c

```diff
@@ -650,3 +700,238 @@ PgCronHasBeenLoaded(void)
+
+	scanDescriptor = systable_beginscan(jobRunDetailTable, indexId, true,
+									NULL, 1, scanKey);
+
+	tupleDescriptor = RelationGetDescr(jobRunDetailTable);
+
+	heapTuple = systable_getnext(scanDescriptor);
+	if (!HeapTupleIsValid(heapTuple))
+	{
+		heapTuple = heap_form_tuple(tupleDescriptor, values, isNulls);
+		#if PG_VERSION_NUM < 100000
+			simple_heap_insert(jobRunDetailTable, heapTuple);
+			CatalogUpdateIndexes(jobRunDetailTable, heapTuple);
+		#else
+			CatalogTupleInsert(jobRunDetailTable, heapTuple);
+		#endif
+		heap_freetuple(heapTuple);
+	}
+	else
+	{
+		HeapTuple newtuple;
+
+		newtuple = heap_modify_tuple(heapTuple, tupleDescriptor, values, isNulls, repl);
+		#if PG_VERSION_NUM < 100000
+			simple_heap_update(jobRunDetailTable, &newtuple->t_self, newtuple);
+			CatalogUpdateIndexes(jobRunDetailTable, newtuple);
+		#else
+			CatalogTupleUpdate(jobRunDetailTable, &newtuple->t_self, newtuple);
+		#endif
+		heap_freetuple(newtuple);
+	}
+
+	systable_endscan(scanDescriptor);
+	table_close(jobRunDetailTable, NoLock);
+
+	PopActiveSnapshot();
+	CommitTransactionCommand();
+
+	return true;
+}
+
+void
+CleanAuditTable(void)
+{
+	Relation jobRunDetailTable;
+	SysScanDesc scanDescriptor;
+	HeapTuple heapTuple;
+	TupleDesc tupleDescriptor;
+	Datum values[Natts_job_run_details];
+	bool isNulls[Natts_job_run_details];
+	bool repl[Natts_job_run_details];
+
+	SetCurrentStatementStartTimestamp();
+	StartTransactionCommand();
+	PushActiveSnapshot(GetTransactionSnapshot());
+
+	/*
+	 * If the pg_job_scheduler extension has not been created yet or
+	 * we are on a hot standby, the job table is treated as
+	 * being empty.
+	 */
+	if (!PgCronHasBeenLoaded() || RecoveryInProgress())
+	{
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+		pgstat_report_activity(STATE_IDLE, NULL);
+		return;
+	}
+
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+	memset(repl, false, sizeof(repl));
+
+	values[Anum_job_run_details_status - 1] = CStringGetTextDatum("failed");
+	repl[Anum_job_run_details_status - 1] = true;
+
+	values[Anum_job_run_details_return_message - 1] = CStringGetTextDatum("server restarted");
+	repl[Anum_job_run_details_return_message - 1] = true;
+
+	get_namespace_oid(CRON_SCHEMA_NAME, false);
+	jobRunDetailTable = table_open(JobRunDetailsRelationId(), RowExclusiveLock);
+
+	scanDescriptor = systable_beginscan(jobRunDetailTable, InvalidOid, false,
+                                                                              NULL, 0, NULL);
+	tupleDescriptor = RelationGetDescr(jobRunDetailTable);
+
+	while (HeapTupleIsValid(heapTuple = systable_getnext(scanDescriptor)))
+	{
+		HeapTuple newtuple;
+		Datum status;
+		bool isNull;
+
+		status = heap_getattr(heapTuple, Anum_job_run_details_status, tupleDescriptor, &isNull);
+
+		if (!isNull &&
+			strcmp("starting", TextDatumGetCString(status)) != 0 &&
+			strcmp("running", TextDatumGetCString(status)) != 0)
+			continue;
```

Comment:
To Jim's point, using SPI would be nicer here. I think we can do that here without having to change everything.